### PR TITLE
[FEATURE] Add encryption

### PR DESCRIPTION
**Pull request by boodyvo**
Add encryption for socket messages for invoice (compatible for receive from an old version of wallet, but cannot send to an old version of wallet)

### CODE REVIEW

File: server/binaries/LocalInvoiceServer.js [hunk 1]
```diff
@@ -19,14 +23,28 @@ const types = {
     SOCKET_UNAUTHORIZED_CONNECTION: "UNAUTHORIZED_CONNECTION",
     SOCKET_SIGN_MESSAGE_SUCCESS: "SIGN_MESSAGE_SUCCESS",
     SOCKET_ADD_INVOICE_REMOTE_RESPONSE: "ADD_INVOICE_REMOTE_RESPONSE",
+    SOCKET_ADD_INVOICE_ENCRYPTED_REMOTE_RESPONSE: "ADD_INVOICE_ENCRYPTED_REMOTE_RESPONSE",
     SOCKET_ADD_INVOICE_REMOTE_REQUEST: "ADD_INVOICE_REMOTE_REQUEST",
```
Comment (serhiislobodian) [inline server/binaries/LocalInvoiceServer.js]: Duplicate with line 17

File: server/binaries/LocalInvoiceServer.js [hunk 2]
```diff
@@ -160,19 +354,37 @@ const openConnection = () => {
 
 const requestInvoice = async (amount, lightningId, memo) => {
     const value = parseInt(amount, 10);
-    const key = `${Date.now()}${crypto.randomBytes(5).toString("hex")}`;
-    invoiceResponse[key] = null;
+    const id = `${Date.now()}${crypto.randomBytes(5).toString("hex")}`;
+
+    // const defParams = {
```
Comment (serhiislobodian) [inline server/binaries/LocalInvoiceServer.js]: Leftoff?

**Comment by lverba** [comment]
retest this please

**Comment by lverba** [comment]
retest this please

**Comment by lverba** [comment]
retest this please

**Comment by lverba** [comment]
retest this please

**Comment by lverba** [comment]
retest this please

File: server/binaries/LocalInvoiceServer.js [hunk 3]
```diff
@@ -19,14 +23,28 @@ const types = {
...
+// here will be keys for encryption, invoice response and memo for invoice
+const invoiceStorage = {};
 const lnd = new Lnd();
 
+// generation of local private key for the wallet session to encrypt communication with other wallets
+const secret = ec.genKeyPair();
+// timout for saving shared keys - 1 hour
+const KEY_SAVE_TIMEOUT = 60 * 60 * 1000;
+// local storage for shared keys
+// let keyStorage = {};
```
Comment (serhiislobodian) [inline server/binaries/LocalInvoiceServer.js]: Something left from dev

File: server/binaries/LocalInvoiceServer.js [hunk 4]
```diff
@@ -127,10 +161,170 @@ const onMessage = async (message) => {
...
+            const checked = await checkAuthority(msg, data.sender);
+            if (!checked.ok) {
+                break;
+            }
+            logger.debug("Checked");
+
+            const pubkey = ec.keyFromPublic(msg.data.pubkey, "hex");
+            // remember shared key for particular id
+
+            // for test
```
Comment (serhiislobodian) [inline server/binaries/LocalInvoiceServer.js]: Left from testing/dev or necessary part?

File: server/binaries/LocalInvoiceServer.js [hunk 5]
```diff
@@ -127,10 +161,170 @@ const onMessage = async (message) => {
...
+            logger.debug("Signing message");
+            const response = await lnd.call("signMessage", { msg: Buffer.from(socketResponse.data.toString()) });
+            socketResponse.sign = response.response.signature;
+
+            logger.debug("Will send:", socketResponse);
+            ws.send(JSON.stringify(socketResponse));
+            break;
+        }
+        case types.SOCKET_PUBKEY_RESPONSE: {
+            logger.debug("Will check auth", invoiceStorage[msg.data.id]);
```
Comment (serhiislobodian) [inline server/binaries/LocalInvoiceServer.js]: Cleanup logger, remove unnecessary, duplicated(for comments too)

**Pull request merged**